File: behavioral-design-patterns/command-design-pattern/Python/uber-rides.py
class RideService:
    def __init__(self) -> None:
        pass
    
    def request_ride(self, passenger: str, src_loc: str, dest_loc: str):
        print(f"Requesting a ride for passenger: {passenger} from {src_loc} to {dest_loc}") 
        # Additional ride request processing logic here
        
    def cancel_ride(self, passenger: str):
        print(f"Cancelling the ride for passenger: {passenger}")
        # Additional Cancellation ride request processing logic here
        
        
class Command:
    def execute(self):
        pass
    
class RideRequestCommand(Command):
    def __init__(self, receiver: RideService, passenger: str, src_loc: str, dest_loc: str) -> None:
        super().__init__()
        self.__receiver = receiver
        self.__passenger = passenger
        self.__src_loc = src_loc
        self.__dest_loc = dest_loc
        
    def execute(self):
        self.__receiver.request_ride(self.__passenger, self.__src_loc, self.__dest_loc)
        
        
class CancelRideRequestCommand(Command):
    def __init__(self, receiver: RideService, passenger: str) -> None:
        super().__init__()
        self.__receiver = receiver
        self.__passenger = passenger
        
    def execute(self):
        self.__receiver.cancel_ride(self.__passenger)
        
class RideRequestInvoker:
    def __init__(self) -> None:
        pass
    
    def process_request(self, command: Command):
        command.execute()

def main():
    # Create a receiver
    ride_service = RideService()
    
    # Create an invoker
    ride_request_invoker = RideRequestInvoker() 
    
    ride_request1: Command = RideRequestCommand(
        receiver=ride_service,   
        passenger="Sumit", 
        src_loc="Mahadevpura", 
        dest_loc="Sarjapur"
    )
    ride_request2: Command = RideRequestCommand(
        receiver=ride_service,
        passenger="Surbhi", 
        src_loc="Sarjapur", 
        dest_loc="Indiranagar"
    )
    cancel_request: Command = CancelRideRequestCommand(
        receiver=ride_service,
        passenger="Sumit"
    )
    
    ride_request_invoker.process_request(ride_request1)
    ride_request_invoker.process_request(ride_request2)
    ride_request_invoker.process_request(cancel_request)
    
       
    
if __name__ == "__main__":
    main()

File: behavioral-design-patterns/command-design-pattern/Typescript/uber-rides.ts
class RideService {
    requestRide(passenger: string, srcLoc: string, destLoc: string){
        console.log(`Requesting a ride for passenger: ${passenger} from ${srcLoc} to ${destLoc}`);
    }

    cancelRide(passenger: string) {
        console.log(`Cancelling the ride for passenger: ${passenger}`);
    }
}

class Command {
    public execute() {}
}

class RideRequestCommand extends Command {
    private receiver: RideService;
    private passenger: string;
    private srcLoc: string;
    private destLoc: string;

    constructor(receiver: RideService, passenger: string, srcLoc: string, destLoc: string) {
        super();
        this.receiver = receiver;
        this.passenger = passenger;
        this.srcLoc = srcLoc;
        this.destLoc = destLoc;
    }

    public execute() {
        this.receiver.requestRide(this.passenger, this.srcLoc, this.destLoc);
    }
}

class CancelRideRequestCommand {
    private receiver: RideService;
    private passenger: string;

    constructor(receiver: RideService, passenger: string) {
        this.receiver = receiver;
        this.passenger = passenger;
    }

    public execute() {
        this.receiver.cancelRide(this.passenger);
    }
}

class RideRequestInvoker {
    public processRequest(command: Command){
        command.execute()
    }
}

function uberRides() {
    const rideService: RideService = new RideService();
    
    const rideRequestInvoker: RideRequestInvoker = new RideRequestInvoker();

    const rideRequest1: Command = new RideRequestCommand(
        rideService,   
        "Sumit", 
        "Mahadevpura", 
        "Sarjapur"
    )
    const rideRequest2: Command = new RideRequestCommand(
        rideService,
        "Surbhi", 
        "Sarjapur", 
        "Indiranagar"
    )
    const cancelRequest: Command = new CancelRideRequestCommand(
        rideService,
        "Sumit"
    )

    rideRequestInvoker.processRequest(rideRequest1);
    rideRequestInvoker.processRequest(rideRequest2);
    rideRequestInvoker.processRequest(cancelRequest);
}

uberRides()